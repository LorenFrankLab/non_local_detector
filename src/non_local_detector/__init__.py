from non_local_detector.models import (
    ContFragClusterlessClassifier,  # noqa
    ContFragSortedSpikesClassifier,  # noqa
    MultiEnvironmentSortedSpikesClassifier,  # noqa
    MultiEnvironmentClusterlessClassifier,  # noqa
    NonLocalClusterlessDetector,  # noqa
    NonLocalSortedSpikesDetector,  # noqa
    ClusterlessDecoder,  # noqa
    SortedSpikesDecoder,  # noqa
)

try:
    from ._version import __version__
except ImportError:
    pass
