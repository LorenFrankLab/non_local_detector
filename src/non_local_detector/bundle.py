"""
non_local_detector.bundle
=========================

* RecordingBundle —— immutable, native-rate archive.
* DecoderBatch     —— time-aligned slice at the decoder's Δt.

All *modalities* live in the open ``signals`` dictionary.
Common ones (`counts`, `lfp`, `calcium`) remain discoverable through
read-only properties for neuroscientist convenience.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, Iterable, List, Optional

import numpy as np

Array = np.ndarray


@dataclass(frozen=True, slots=True)
class SpikeTrain:
    times_s: np.ndarray  # strictly increasing
    unit_id: int | str  # cluster or channel label
    channel_position: tuple[float, ...] | None = None  # (x, y, z) if known
    quality_metrics: dict[str, float] = field(default_factory=dict)


@dataclass(frozen=True, slots=True)
class WaveformSeries:
    data: np.ndarray  # (n_spikes, n_channels, n_samples)
    channel_positions: np.ndarray | None = None  # (n_channels, ndim)
    channel_ids: tuple[int, ...] | None = None
    feature_names: tuple[str, ...] | None = None  # e.g. ("amp", "width")


# --------------------------------------------------------------------------- #
#  Helper: generic time-series container                                      #
# --------------------------------------------------------------------------- #
@dataclass(frozen=True, slots=True)
class TimeSeries:
    data: Array  # shape (n_samples, …)
    sampling_rate_hz: float  # samples per second
    start_s: float = 0.0  # session time of first sample

    # OPTIONAL metadata  ──────────────────────────────────────────────
    channel_ids: tuple[int, ...] | None = None  # length = n_channels
    channel_positions: np.ndarray | None = None
    units: str | None = None  # "uV", "raw_adc", "deg"

    def __post_init__(self):
        if self.sampling_rate_hz <= 0:
            raise ValueError("sampling_rate_hz must be positive.")
        if self.data.ndim < 1:
            raise ValueError("TimeSeries.data must be at least 1-D.")


# --------------------------------------------------------------------------- #
#  Native-rate archive                                                        #
# --------------------------------------------------------------------------- #
@dataclass(frozen=True, slots=True)
class RecordingBundle:
    """
    Immutable store for raw session data at native sampling rates.
    """

    # List of spike times for each neuron, in seconds.
    spike_times_s: Optional[List[SpikeTrain]] = None
    # List of spike waveforms, one for each neuron.
    spike_waveforms: Optional[List[WaveformSeries]] = None
    signals: Dict[str, TimeSeries] = field(default_factory=dict)

    def __post_init__(self) -> None:
        # ---- 1. waveforms without timestamps → fatal -----------------
        if self.spike_waveforms is not None and self.spike_times_s is None:
            raise ValueError(
                "spike_waveforms were provided but spike_times_s is None. "
                "Waveforms cannot be aligned without spike times."
            )

        # ---- 2. no timestamps → nothing else to check ----------------
        if self.spike_times_s is None:
            return  # (case 1 above)

        # ---- 3. timestamps present, waveforms missing → fine ---------
        if self.spike_waveforms is None:
            return  # (case 2 above)

        # ---- 4. both present → validate lengths & per-unit matches ---
        if len(self.spike_times_s) != len(self.spike_waveforms):
            raise ValueError(
                "Length mismatch between spike_times_s "
                f"({len(self.spike_times_s)}) and spike_waveforms "
                f"({len(self.spike_waveforms)})."
            )

        for idx, (t, w) in enumerate(zip(self.spike_times_s, self.spike_waveforms)):
            ntimes = t.times_s.size
            nwaves = w.data.shape[0]
            if ntimes != nwaves:
                raise ValueError(
                    f"Neuron {idx}: {ntimes} spike times but "
                    f"{nwaves} waveform rows."
                )


# --------------------------------------------------------------------------- #
#  Decoder-aligned batch                                                      #
# --------------------------------------------------------------------------- #
@dataclass(slots=True)
class DecoderBatch:
    """
    Time-aligned view at the decoder's bin width Δt (e.g. 50 Hz).
    Every array in ``signals`` **must** share the first-axis length ``n_time``.
    signals can be any time-series modality, such as:
    - `counts` (spike counts, shape (n_time,))
    - `lfp` (local field potential, shape (n_time, n_channels))
    - `calcium` (fluorescence, shape (n_time, n_channels))
    Optionally, it can also contain raw spike times and waveforms.

    Attributes
    ----------
    signals : Dict[str, Array]
        Dictionary of time-series arrays, where each key is a modality name
        and the value is the corresponding data array.
    bin_edges_s : Optional[Array]
        Optional array of bin edges in seconds, with length equal to `n_time + 1`.
        If provided, it defines the time intervals for each bin.
    spike_times_s : Optional[List[Array]]
        Optional list of spike times for each neuron, where each array
        corresponds to a different neuron and contains spike times in seconds.
    spike_waveforms : Optional[List[Array]]
        Optional list of spike waveforms, where each array corresponds to
        a different neuron and contains the waveforms of spikes.
    """

    signals: Dict[str, Array] = field(default_factory=dict)

    # Per-bin metadata
    bin_edges_s: Optional[Array] = None  # len = n_time + 1

    # Raw spikes
    spike_times_s: Optional[List[Array]] = None
    spike_waveforms: Optional[List[Array]] = None

    _n_time: int = field(init=False, repr=False)

    # ------------------------------------------------------------------ #
    #  Validation at construction                                        #
    # ------------------------------------------------------------------ #
    def __post_init__(self):
        # 1) Neither signals nor spikes → error
        if not self.signals and self.spike_times_s is None:
            raise ValueError(
                "DecoderBatch must contain at least one signal or spike_times_s."
            )

        if self.signals:
            # a) Disallow any non-ndarray
            bad_type = [
                k for k, v in self.signals.items() if not isinstance(v, np.ndarray)
            ]
            if bad_type:
                raise TypeError(f"signals must be np.ndarray; offenders: {bad_type}")

            # b) Now each v is an ndarray → check dtype
            bad_dtype = [
                k for k, v in self.signals.items() if v.dtype.kind not in "fi?uO"
            ]
            if bad_dtype:
                raise TypeError(
                    f"signals must be numeric or bool; offenders: {bad_dtype}"
                )

            # c) Ensure all signals share the same n_time
            lengths = {k: v.shape[0] for k, v in self.signals.items()}
            if len(set(lengths.values())) != 1:
                detail = ", ".join(f"{k}={n}" for k, n in lengths.items())
                raise ValueError(f"Time-axis mismatch among signals: {detail}")
            self._n_time = next(iter(lengths.values()))

        else:
            # No signals, but spike_times_s is not None (otherwise we would have raised above).
            # We allow “spikes + bin_edges only” → set n_time from bin_edges_s.
            if self.spike_times_s is not None and self.bin_edges_s is not None:
                self._n_time = len(self.bin_edges_s) - 1
            else:
                # If spike_times_s exists but bin_edges_s is missing, we cannot infer n_time.
                raise ValueError(
                    "DecoderBatch must contain either signals or both spike_times_s and bin_edges_s."
                )

        # Final check on bin_edges_s length (if provided)
        if self.bin_edges_s is not None:
            if len(self.bin_edges_s) != self._n_time + 1:
                raise ValueError("len(bin_edges_s) must equal n_time + 1.")

    # ------------------------------------------------------------------ #
    #  Convenience read-only attributes                                  #
    # ------------------------------------------------------------------ #
    @property
    def counts(self) -> Optional[Array]:
        return self.signals.get("counts")

    @property
    def lfp(self) -> Optional[Array]:
        return self.signals.get("lfp")

    @property
    def calcium(self) -> Optional[Array]:
        return self.signals.get("calcium")

    @property
    def n_time(self) -> int:
        return self._n_time

    # ------------------------------------------------------------------ #
    #  Shallow slice helper                                              #
    # ------------------------------------------------------------------ #
    def slice(
        self, start: int, stop: int, *, slice_spikes: bool = True
    ) -> "DecoderBatch":
        new_signals = {k: v[start:stop] for k, v in self.signals.items()}

        st, wf = self.spike_times_s, self.spike_waveforms
        if slice_spikes and st is not None and self.bin_edges_s is not None:
            t0, t1 = self.bin_edges_s[start], self.bin_edges_s[stop]
            st_new = []
            wf_new = [] if wf is not None else None

            for i, times in enumerate(st):
                mask = (times >= t0) & (times < t1)
                st_new.append(times[mask])
                if wf is not None:
                    wf_new.append(wf[i][mask])

            st, wf = st_new, (wf_new if wf is not None else None)

        new_edges = (
            None if self.bin_edges_s is None else self.bin_edges_s[start : stop + 1]
        )
        return DecoderBatch(
            signals=new_signals,
            bin_edges_s=new_edges,
            spike_times_s=st,
            spike_waveforms=wf,
        )

    def select_signals(self, keys: Iterable[str]) -> "DecoderBatch":
        return DecoderBatch(
            signals={k: self.signals[k] for k in keys},
            bin_edges_s=self.bin_edges_s,
            spike_times_s=self.spike_times_s,
            spike_waveforms=self.spike_waveforms,
        )

    def select_spikes(self, keys: Iterable[int]) -> "DecoderBatch":
        if self.spike_times_s is None:
            raise ValueError("No spike_times_s available in this DecoderBatch.")

        orig_st = self.spike_times_s
        orig_wf = self.spike_waveforms

        new_times = []
        for i in keys:
            t = orig_st[i]
            if hasattr(t, "times_s"):
                new_times.append(t.times_s)
            else:
                new_times.append(t)

        if orig_wf is not None:
            new_wf = []
            for i in keys:
                w = orig_wf[i]
                if hasattr(w, "data"):
                    new_wf.append(w.data)
                else:
                    new_wf.append(w)
        else:
            new_wf = None

        return DecoderBatch(
            signals=self.signals,
            bin_edges_s=self.bin_edges_s,
            spike_times_s=new_times,
            spike_waveforms=new_wf,
        )

    # ------------------------------------------------------------------ #
    #  Pretty-print                                                      #
    # ------------------------------------------------------------------ #
    def __repr__(self) -> str:  # pragma: no cover
        sigs = ", ".join(self.signals.keys())
        return f"<DecoderBatch n_time={self.n_time} signals=[{sigs}]>"


# --------------------------------------------------------------------------- #
#  Public helper: bundle-level validation                                     #
# --------------------------------------------------------------------------- #
def validate_sources(batch: "DecoderBatch", observation_models: list) -> None:
    """
    Check that every ObservationModel's ``required_sources`` are
    present in *this* batch.

    Usage
    -----
    >>> validate_sources(batch, detector.observation_models)
    """
    missing: set[str] = set()
    for model in observation_models:
        for src in getattr(model, "required_sources", ()):
            # OK if it’s in signals dict
            if src in batch.signals:
                continue
            # OK if there is an attribute with that name AND it’s not None
            if hasattr(batch, src) and getattr(batch, src) is not None:
                continue
            # Otherwise, it’s missing
            missing.add(src)
    if missing:
        raise ValueError(
            "DecoderBatch missing required fields:\n  • "
            + "\n  • ".join(sorted(missing))
        )
