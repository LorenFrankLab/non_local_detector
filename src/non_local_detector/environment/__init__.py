from non_local_detector.environment.alignment import (
    get_2d_rotation_matrix,
    map_probabilities_to_nearest_target_bin,
)
from non_local_detector.environment.environment import Environment
from non_local_detector.environment.layout.factories import (
    get_layout_parameters,
    list_available_layouts,
)

__all__ = [
    "get_2d_rotation_matrix",
    "map_probabilities_to_nearest_target_bin",
    "Environment",
    "get_layout_parameters",
    "list_available_layouts",
]
