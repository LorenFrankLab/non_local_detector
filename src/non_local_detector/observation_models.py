"""Class for representing an environment and a condition (trial type, etc.)"""

from dataclasses import dataclass


@dataclass(order=True)
class ObservationModel:
    """Determines which environment and data points data correspond to.

    Attributes
    ----------
    environment_name : str, optional
    encoding_group : str, optional
    is_local : bool, optional
    is_no_spike : bool, optional

    """

    environment_name: str = ""
    encoding_group: str = 0
    is_local: bool = False
    is_no_spike: bool = False

    def __eq__(self, other):
        if other.__class__ is not self.__class__:
            return NotImplemented
        return (self.environment_name, self.encoding_group) == (
            other.environment_name,
            other.encoding_group,
        )
