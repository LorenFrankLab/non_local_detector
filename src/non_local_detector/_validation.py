"""Internal validation utilities for parameter checking.

This module provides reusable validation functions following Raymond Hettinger's
principle: "Functions are better than classes when you don't need state."

These functions are for internal use only (note the leading underscore in module name).
"""

from typing import Any

import numpy as np

from non_local_detector.exceptions import DataError, ValidationError


def ensure_probability_distribution(
    arr: np.ndarray, name: str, tolerance: float = 1e-6
) -> None:
    """Verify array is a valid probability distribution (sums to 1).

    Parameters
    ----------
    arr : np.ndarray
        Array that should sum to 1.0
    name : str
        Name of the array for error messages
    tolerance : float, optional
        Tolerance for sum check, by default 1e-6

    Raises
    ------
    ValidationError
        If array does not sum to 1.0 within tolerance

    Examples
    --------
    >>> ensure_probability_distribution(np.array([0.5, 0.5]), "probs")  # OK
    >>> ensure_probability_distribution(np.array([0.6, 0.6]), "probs")  # Raises
    """
    arr_sum = np.sum(arr)
    if not np.isclose(arr_sum, 1.0, atol=tolerance):
        raise ValidationError(
            f"{name} must be a valid probability distribution",
            expected="sum = 1.0",
            got=f"sum = {arr_sum:.6f}",
            hint=f"Normalize the array: {name} = {name} / {name}.sum()",
            example=f"    {name} = np.array([0.25, 0.25, 0.25, 0.25])  # sums to 1.0",
        )


def ensure_positive_scalar(
    value: float, name: str, minimum: float = 0.0, strict: bool = True
) -> None:
    """Verify value is positive (or non-negative).

    Parameters
    ----------
    value : float
        Value to check
    name : str
        Name of the parameter for error messages
    minimum : float, optional
        Minimum allowed value, by default 0.0
    strict : bool, optional
        If True, value must be > minimum. If False, value must be >= minimum.
        By default True.

    Raises
    ------
    ValidationError
        If value does not meet the constraint

    Examples
    --------
    >>> ensure_positive_scalar(0.5, "concentration")  # OK
    >>> ensure_positive_scalar(0.0, "concentration")  # Raises (strict=True)
    >>> ensure_positive_scalar(0.0, "regularization", strict=False)  # OK
    """
    if strict:
        condition = value > minimum
        expected_str = f"{name} > {minimum}"
    else:
        condition = value >= minimum
        expected_str = f"{name} >= {minimum}"

    if not condition:
        raise ValidationError(
            f"Invalid value for {name}",
            expected=expected_str,
            got=f"{name} = {value}",
            hint=f"Use a {'positive' if strict else 'non-negative'} value",
            example=f"    {name} = {1.0 if strict else 0.0}",
        )


def ensure_array_1d(arr: np.ndarray, name: str) -> None:
    """Verify array is 1-dimensional.

    Parameters
    ----------
    arr : np.ndarray
        Array to check
    name : str
        Name of the array for error messages

    Raises
    ------
    ValidationError
        If array is not 1D

    Examples
    --------
    >>> ensure_array_1d(np.array([1, 2, 3]), "probs")  # OK
    >>> ensure_array_1d(np.array([[1, 2]]), "probs")  # Raises
    """
    if arr.ndim != 1:
        raise ValidationError(
            f"{name} must be a 1-dimensional array",
            expected="array with shape (n,)",
            got=f"array with shape {arr.shape}",
            hint="Use array.flatten() or array.ravel() to convert to 1D",
            example=f"    {name} = np.array([0.25, 0.25, 0.25, 0.25])",
        )


def ensure_all_finite(arr: np.ndarray, name: str) -> None:
    """Verify all array elements are finite (no NaN or Inf).

    Parameters
    ----------
    arr : np.ndarray
        Array to check
    name : str
        Name of the array for error messages

    Raises
    ------
    DataError
        If array contains NaN or Inf values

    Examples
    --------
    >>> ensure_all_finite(np.array([1.0, 2.0]), "data")  # OK
    >>> ensure_all_finite(np.array([1.0, np.nan]), "data")  # Raises
    """
    if not np.all(np.isfinite(arr)):
        n_nan = np.sum(np.isnan(arr))
        n_inf = np.sum(np.isinf(arr))
        raise DataError(
            f"Found non-finite values in {name}",
            data_name=name,
            hint=f"Array contains {n_nan} NaN value(s) and {n_inf} Inf value(s). Check your data for missing or invalid values.",
        )


def ensure_all_non_negative(arr: np.ndarray, name: str) -> None:
    """Verify all array elements are non-negative.

    Parameters
    ----------
    arr : np.ndarray
        Array to check
    name : str
        Name of the array for error messages

    Raises
    ------
    ValidationError
        If array contains negative values

    Examples
    --------
    >>> ensure_all_non_negative(np.array([0.0, 1.0]), "probs")  # OK
    >>> ensure_all_non_negative(np.array([-0.1, 1.0]), "probs")  # Raises
    """
    if np.any(arr < 0):
        n_negative = np.sum(arr < 0)
        min_val = np.min(arr)
        raise ValidationError(
            f"{name} must contain only non-negative values",
            expected="all values >= 0",
            got=f"{n_negative} negative value(s), minimum = {min_val:.6f}",
            hint="Check your data for errors or use np.abs() if negative values should be positive",
        )


def ensure_in_range(arr: np.ndarray, name: str, low: float, high: float) -> None:
    """Verify all array elements are within a range [low, high].

    Parameters
    ----------
    arr : np.ndarray
        Array to check
    name : str
        Name of the array for error messages
    low : float
        Minimum allowed value (inclusive)
    high : float
        Maximum allowed value (inclusive)

    Raises
    ------
    ValidationError
        If array contains values outside the range

    Examples
    --------
    >>> ensure_in_range(np.array([0.5, 0.8]), "probs", 0.0, 1.0)  # OK
    >>> ensure_in_range(np.array([1.5]), "probs", 0.0, 1.0)  # Raises
    """
    if np.any(arr < low) or np.any(arr > high):
        min_val = np.min(arr)
        max_val = np.max(arr)
        raise ValidationError(
            f"{name} values must be in range [{low}, {high}]",
            expected=f"all values in [{low}, {high}]",
            got=f"values in [{min_val:.6f}, {max_val:.6f}]",
            hint=f"Use np.clip({name}, {low}, {high}) to clamp values to valid range",
        )


def ensure_square_matrix(matrix: np.ndarray, name: str) -> None:
    """Verify matrix is square (n x n).

    Parameters
    ----------
    matrix : np.ndarray
        Matrix to check
    name : str
        Name of the matrix for error messages

    Raises
    ------
    ValidationError
        If matrix is not square

    Examples
    --------
    >>> ensure_square_matrix(np.eye(3), "transition_matrix")  # OK
    >>> ensure_square_matrix(np.zeros((3, 4)), "transition_matrix")  # Raises
    """
    if matrix.ndim != 2:
        raise ValidationError(
            f"{name} must be a 2-dimensional matrix",
            expected="matrix with shape (n, n)",
            got=f"array with shape {matrix.shape}",
            hint="Transition matrices must be 2D and square",
        )

    if matrix.shape[0] != matrix.shape[1]:
        raise ValidationError(
            f"{name} must be square",
            expected=f"matrix with shape ({matrix.shape[0]}, {matrix.shape[0]})",
            got=f"matrix with shape {matrix.shape}",
            hint="Transition matrix needs one row and one column per state",
        )


def ensure_stochastic_matrix(
    matrix: np.ndarray, name: str, tolerance: float = 1e-6
) -> None:
    """Verify matrix is row-stochastic (each row sums to 1).

    Parameters
    ----------
    matrix : np.ndarray
        Matrix to check
    name : str
        Name of the matrix for error messages
    tolerance : float, optional
        Tolerance for row sum check, by default 1e-6

    Raises
    ------
    ValidationError
        If any row does not sum to 1.0 within tolerance

    Examples
    --------
    >>> ensure_stochastic_matrix(np.eye(3), "transition_matrix")  # OK
    >>> bad_matrix = np.array([[0.5, 0.5], [0.6, 0.3]])  # Row 2 sums to 0.9
    >>> ensure_stochastic_matrix(bad_matrix, "transition_matrix")  # Raises
    """
    row_sums = np.sum(matrix, axis=1)
    bad_rows = ~np.isclose(row_sums, 1.0, atol=tolerance)

    if np.any(bad_rows):
        bad_indices = np.where(bad_rows)[0]
        bad_sums = row_sums[bad_rows]
        raise ValidationError(
            f"{name} must be row-stochastic (each row sums to 1)",
            expected="all row sums = 1.0",
            got=f"{len(bad_indices)} row(s) with invalid sums: rows {bad_indices.tolist()}, sums {bad_sums.tolist()}",
            hint=f"Normalize each row: {name} = {name} / {name}.sum(axis=1, keepdims=True)",
            example=f"    # Row 0 sums to {row_sums[0]:.6f}{'  âœ“' if not bad_rows[0] else '  âœ—'}",
        )


def ensure_ndarray(value: Any, name: str) -> None:
    """Verify value is a numpy array.

    Parameters
    ----------
    value : Any
        Value to check
    name : str
        Name of the parameter for error messages

    Raises
    ------
    ValidationError
        If value is not a numpy array

    Examples
    --------
    >>> ensure_ndarray(np.array([1, 2]), "position")  # OK
    >>> ensure_ndarray([1, 2], "position")  # Raises
    """
    if not isinstance(value, np.ndarray):
        raise ValidationError(
            f"{name} must be a numpy array",
            expected="np.ndarray",
            got=f"{type(value).__name__}",
            hint=f"Convert to numpy array: {name} = np.array({name})",
            example=f"    {name} = np.array([1.0, 2.0, 3.0])",
        )


def ensure_monotonic_increasing(
    arr: np.ndarray, name: str, strict: bool = False
) -> None:
    """Verify array is monotonically increasing.

    Parameters
    ----------
    arr : np.ndarray
        Array to check
    name : str
        Name of the array for error messages
    strict : bool, optional
        If True, require strictly increasing (no equal consecutive values).
        By default False.

    Raises
    ------
    DataError
        If array is not monotonically increasing

    Examples
    --------
    >>> ensure_monotonic_increasing(np.array([1, 2, 3]), "time")  # OK
    >>> ensure_monotonic_increasing(np.array([1, 3, 2]), "time")  # Raises
    """
    diffs = np.diff(arr)

    if strict:
        condition = np.all(diffs > 0)
        expected_str = "strictly increasing (each value > previous)"
    else:
        condition = np.all(diffs >= 0)
        expected_str = "monotonically increasing (each value >= previous)"

    if not condition:
        # Find first violation
        if strict:
            bad_indices = np.where(diffs <= 0)[0]
        else:
            bad_indices = np.where(diffs < 0)[0]

        if len(bad_indices) > 0:
            idx = bad_indices[0]
            raise DataError(
                f"{name} must be {expected_str}",
                data_name=name,
                hint=f"Found violation at index {idx}: {name}[{idx}] = {arr[idx]:.6f}, {name}[{idx + 1}] = {arr[idx + 1]:.6f}. Check for duplicate or out-of-order timestamps.",
            )


def ensure_matching_lengths(arr1, arr2, name1: str, name2: str) -> None:
    """Verify two arrays have matching lengths.

    Parameters
    ----------
    arr1 : array-like
        First array
    arr2 : array-like
        Second array
    name1 : str
        Name of first array for error messages
    name2 : str
        Name of second array for error messages

    Raises
    ------
    ValidationError
        If arrays have different lengths

    Examples
    --------
    >>> ensure_matching_lengths(np.array([1, 2, 3]), np.array([4, 5, 6]), "pos", "time")  # OK
    >>> ensure_matching_lengths(np.array([1, 2]), np.array([4, 5, 6]), "pos", "time")  # Raises
    """
    len1 = len(arr1)
    len2 = len(arr2)

    if len1 != len2:
        raise ValidationError(
            f"Length mismatch between {name1} and {name2}",
            expected=f"{name2} with length {len1}",
            got=f"{name2} with length {len2}",
            hint=f"Ensure {name1} and {name2} have the same number of time points",
            example=f"    # Both should have same length\n    len({name1}) = {len1}\n    len({name2}) = {len2}  # Mismatch!",
        )
