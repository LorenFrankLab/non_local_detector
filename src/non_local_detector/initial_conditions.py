from dataclasses import dataclass

import numpy as np

from non_local_detector.environment import Environment
from non_local_detector.observation_models import ObservationModel


@dataclass
class UniformInitialConditions:
    """Initial conditions where all position bins are
    equally likely."""

    def make_initial_conditions(
        self,
        observation_model: ObservationModel,
        environments: list[Environment],
    ) -> np.ndarray:
        """Creates initial conditions array

        Parameters
        ----------
        observation_model : ObservationModel
        environments : list[Environment]

        Returns
        -------
        unnormalized_initial_conditions : np.ndarray, shape (n_place_bins,)
        """

        if observation_model.is_local or observation_model.is_no_spike:
            initial_conditions = np.ones((1,), dtype=np.float32)
        else:
            environment = environments[
                environments.index(observation_model.environment_name)
            ]
            initial_conditions = environment.is_track_interior_.ravel().astype(
                np.float32
            )

        initial_conditions /= initial_conditions.sum()

        return initial_conditions


def estimate_initial_conditions(acausal_posterior: np.ndarray) -> np.ndarray:
    """Estimate initial conditions from acausal posterior distribution via EM algorithm

    Parameters
    ----------
    acausal_posterior : np.ndarray, shape (n_time, n_states)
        Acausal posterior distribution

    Returns
    -------
    initial_conditions : np.ndarray, shape (n_states,)
        Estimated initial conditions

    """
    return acausal_posterior[0]
