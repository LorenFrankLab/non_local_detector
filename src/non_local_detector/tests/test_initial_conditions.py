import numpy as np
import pytest

from non_local_detector.environment import Environment
from non_local_detector.initial_conditions import (
    UniformInitialConditions,
    estimate_initial_conditions,
)
from non_local_detector.observation_models import ObservationModel

from non_local_detector.tests.conftest import assert_probability_distribution


@pytest.fixture
def make_env_1d():
    """Factory fixture for creating 1D environments with custom parameters."""
    def _make_env(n_bins=11):
        env = Environment(
            environment_name="line",
            place_bin_size=1.0,
            position_range=((0.0, float(n_bins - 1)),),
        )
        pos = np.linspace(0.0, float(n_bins - 1), n_bins)[:, None]
        env = env.fit_place_grid(position=pos, infer_track_interior=False)
        return env
    return _make_env


def test_uniform_initial_conditions_local_and_no_spike(make_env_1d):
    ic = UniformInitialConditions()
    envs = (make_env_1d(),)

    # Local decoding => single-bin initial conditions
    local_obs = ObservationModel(
        environment_name=envs[0].environment_name, is_local=True
    )
    arr = ic.make_initial_conditions(local_obs, envs)
    assert arr.shape == (1,)
    assert np.isclose(arr.sum(), 1.0) and np.isclose(arr[0], 1.0)

    # No-spike => single-bin initial conditions
    ns_obs = ObservationModel(
        environment_name=envs[0].environment_name, is_no_spike=True
    )
    arr2 = ic.make_initial_conditions(ns_obs, envs)
    assert arr2.shape == (1,) and np.isclose(arr2[0], 1.0)


def test_uniform_initial_conditions_nonlocal_with_mask(make_env_1d):
    env = make_env_1d(n_bins=10)
    # Mask out last two bins
    mask = np.ones(env.centers_shape_, dtype=bool)
    mask[-2:] = False
    env.is_track_interior_ = mask
    envs = (env,)
    ic = UniformInitialConditions()
    obs = ObservationModel(environment_name=env.environment_name, is_local=False)
    arr = ic.make_initial_conditions(obs, envs)
    assert arr.shape == (env.place_bin_centers_.shape[0],)
    assert np.isclose(arr.sum(), 1.0)
    # Zero probability on masked bins
    assert np.all(arr[-2:] == 0.0)
    # Uniform across unmasked bins
    assert np.allclose(arr[:-2], 1.0 / (env.place_bin_centers_.shape[0] - 2))


def test_estimate_initial_conditions_returns_first_row():
    post = np.random.default_rng(0).random((7, 5))
    post = post / post.sum(axis=1, keepdims=True)
    init = estimate_initial_conditions(post)
    assert init.shape == (post.shape[1],)
    assert np.allclose(init, post[0])


def test_uniform_initial_conditions_no_track_interior(make_env_1d):
    """Test when environment has no track interior mask."""
    env = make_env_1d(n_bins=5)
    env.is_track_interior_ = None  # No mask
    envs = (env,)
    ic = UniformInitialConditions()
    obs = ObservationModel(environment_name=env.environment_name, is_local=False)

    # Act
    arr = ic.make_initial_conditions(obs, envs)

    # Assert - should fallback to single bin
    assert arr.shape == (1,)
    assert np.isclose(arr[0], 1.0)


def test_uniform_initial_conditions_normalization(make_env_1d):
    """Test that initial conditions always sum to 1."""
    env = make_env_1d(n_bins=20)
    envs = (env,)
    ic = UniformInitialConditions()
    obs = ObservationModel(environment_name=env.environment_name, is_local=False)

    # Act
    arr = ic.make_initial_conditions(obs, envs)

    # Assert
    assert_probability_distribution(arr)


def test_estimate_initial_conditions_preserves_distribution():
    """Estimate should preserve the probability distribution at t=0."""
    # Arrange - create a non-uniform initial distribution
    post = np.array([[0.7, 0.2, 0.1], [0.5, 0.3, 0.2], [0.4, 0.4, 0.2]])

    # Act
    init = estimate_initial_conditions(post)

    # Assert
    assert_probability_distribution(init)
    assert np.array_equal(init, post[0])
