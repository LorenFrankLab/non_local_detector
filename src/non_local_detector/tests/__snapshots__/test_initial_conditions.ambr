# serializer version: 1
# name: test_estimate_initial_conditions_snapshot
  dict({
    'dtype': 'float64',
    'max': 0.3707334390979684,
    'mean': 0.12499999999999997,
    'min': 0.007369879583536177,
    'shape': tuple(
      8,
    ),
    'std': 0.12094052461209008,
    'sum': 0.9999999999999998,
    'values': list([
      0.0577961311026003,
      0.3707334390979684,
      0.16217341080381148,
      0.11244009267661835,
      0.020891386951085476,
      0.020887867151791112,
      0.007369879583536177,
      0.24770779263258855,
    ]),
  })
# ---
# name: test_uniform_initial_conditions_local_snapshot
  dict({
    'dtype': 'float32',
    'max': 1.0,
    'mean': 1.0,
    'min': 1.0,
    'shape': tuple(
      1,
    ),
    'std': 0.0,
    'sum': 1.0,
    'values': list([
      1.0,
    ]),
  })
# ---
# name: test_uniform_initial_conditions_nonlocal_snapshot
  dict({
    'dtype': 'float32',
    'max': 0.0714285746216774,
    'mean': 0.071428582072258,
    'min': 0.0714285746216774,
    'shape': tuple(
      14,
    ),
    'std': 7.450580596923828e-09,
    'sum': 1.0000001192092896,
    'values': list([
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
      0.0714285746216774,
    ]),
  })
# ---
# name: test_uniform_initial_conditions_with_mask_snapshot
  dict({
    'dtype': 'float32',
    'max': 0.1666666716337204,
    'mean': 0.09090909361839294,
    'min': 0.0,
    'shape': tuple(
      11,
    ),
    'std': 0.08298826962709427,
    'sum': 1.0,
    'values': list([
      0.1666666716337204,
      0.1666666716337204,
      0.0,
      0.0,
      0.1666666716337204,
      0.1666666716337204,
      0.1666666716337204,
      0.1666666716337204,
      0.0,
      0.0,
      0.0,
    ]),
  })
# ---
